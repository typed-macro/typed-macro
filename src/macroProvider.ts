import {
  NamespacedModules,
  NamespacedMacros,
  NamespacedTypes,
} from '@/runtime/types'
import { ResolvedConfig, ViteDevServer } from 'vite'
import { DevServerHelper } from '@/helper/server'

export type MacroProviderHooks = {
  onViteStart?: (
    config: ResolvedConfig,
    server: ViteDevServer,
    helper: DevServerHelper
  ) => void | Promise<void>
  onRollupStart?: () => void | Promise<void>
  onStart?: () => void | Promise<void>
}

export type MacroProvider = {
  id: string
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
  hooks: MacroProviderHooks
}

interface InternalMacroProvider extends MacroProvider {
  __internal_macro_provider: true
}

export function isMacroProvider(o: unknown): o is MacroProvider {
  return (o as any).__internal_macro_provider
}

export function macroProvider(provider: MacroProvider): MacroProvider {
  ;(provider as InternalMacroProvider).__internal_macro_provider = true
  return provider
}
