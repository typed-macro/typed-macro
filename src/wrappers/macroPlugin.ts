import type { Plugin, ViteDevServer } from 'vite'
import { DevServerHelper, getDevServerHelper } from '@/wrappers/helper/server'
import { Attachable, Runtime } from '@/core/runtime'
import { versionedPlugin } from '@/wrappers/compat'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtime: Runtime
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => Attachable
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = options.runtime

  const plugin = {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const consume = {
        exports: runtime.exports,
        options: runtime.options,
      }
      runtime = undefined
      return consume
    },
    name,
    enforce: 'pre',
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    configureServer(server) {
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.call(this, opt),
      ])
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.call(this, id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.call(this, id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return (
        transform?.call(this, transformed ?? code, id, ssr) ??
        (transformed && { code: transformed, map: null })
      )
    },
    ...otherHooks,
  } as InternalMacroPlugin
  return versionedPlugin(plugin)
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any)?.__internal_macro_plugin ?? false
}
