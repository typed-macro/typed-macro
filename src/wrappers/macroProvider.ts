import { NormalizedExports } from '@/core/exports'
import { ResolvedConfig, ViteDevServer } from 'vite'
import { DevServerHelper } from '@/wrappers/helper/server'
import { RuntimeOptions } from '@/core/runtime'

export type ViteStartContext = (
  | {
      /**
       * Is in dev mode.
       */
      dev: true
      /**
       * Vite dev server.
       */
      server: ViteDevServer
      /**
       * Wrappers on vite dev server.
       */
      helper: DevServerHelper
    }
  | {
      /**
       * Is in dev mode.
       */
      dev: false
    }
) & {
  /**
   * Vite resolved config.
   */
  config: ResolvedConfig
}

export type MacroProviderHooks = {
  /**
   * A startup hook called only if running in Vite.
   */
  onViteStart?: (ctx: ViteStartContext) => void | Promise<void>
  /**
   * A startup hook called only if running in Rollup.
   */
  onRollupStart?: () => void | Promise<void>
  /**
   * A startup hook, called immediately after {@link onViteStart}/{@link onRollupStart}.
   */
  onStart?: () => void | Promise<void>
}

export type MacroProvider = {
  id: string
  exports: NormalizedExports
  hooks: MacroProviderHooks
  options?: Partial<Pick<RuntimeOptions, 'transformer'>>
}

interface InternalMacroProvider extends MacroProvider {
  __internal_macro_provider: true
}

export function isMacroProvider(o: unknown): o is MacroProvider {
  return (o as any)?.__internal_macro_provider ?? false
}

export function macroProvider(provider: MacroProvider): MacroProvider {
  ;(provider as InternalMacroProvider).__internal_macro_provider = true
  return provider
}
