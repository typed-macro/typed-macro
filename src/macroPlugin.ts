import type { Plugin, ViteDevServer } from 'vite'
import { NormalizedExports } from '@/core/exports'
import { MacroProvider } from '@/macroProvider'
import { DevServerHelper, getDevServerHelper } from '@/helper/server'
import { Runtime, RuntimeOptions } from '@/core/runtime'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtimeOptions: RuntimeOptions
  exports: NormalizedExports
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => MacroProvider
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    exports,
    runtimeOptions,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = new Runtime(runtimeOptions)

  runtime.register(exports)

  return {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const provider: MacroProvider = {
        id: name,
        exports: runtime.exports,
        hooks: {},
      }
      runtime = undefined
      return provider
    },
    name,
    enforce: 'pre',
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.bind(this)(opt),
      ])
    },
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.bind(this)(id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.bind(this)(id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return transform?.bind(this)(transformed ?? code, id, ssr) ?? transformed
    },
    configureServer(server) {
      // hook
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    ...otherHooks,
  } as InternalMacroPlugin
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any).__internal_macro_plugin ?? false
}
