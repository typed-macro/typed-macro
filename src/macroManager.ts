import { Plugin, ResolvedConfig, ViteDevServer } from 'vite'
import {
  isMacroProvider,
  MacroProvider,
  MacroProviderHooks,
} from '@/macroProvider'
import { isMacroPlugin, macroPlugin, MacroPlugin } from '@/macroPlugin'
import { getDevServerHelper } from '@/helper/server'
import { Runtime } from '@/core/runtime'

/**
 * A manager/container of macros.
 */
export type MacroManager = {
  /**
   * Register {@link MacroProvider} or {@link MacroPlugin} to this macro manager so that
   * all macros in providers and plugins share the same runtime.
   *
   * For macro plugins:
   *  > Some options like `maxRecursions` or `typesPath` will be overridden by
   *  > manager's, `parserPlugins` will be merged with the manager's one.
   *  >
   *  > After registered, the original macro plugin will be attached to the manager,
   *  > which means no need to add the plugin to Vite/Rollup 's plugins array again.
   * @param sources macro providers or plugins.
   */
  use(...sources: (MacroProvider | Plugin)[]): MacroManager
  /**
   * Get a shallow copy of the underlying plugins of the macro manager.
   * In most cases, it can be used as one plugin directly since Vite flats
   * plugin array automatically.
   */
  toPlugin(): Plugin[]
}

export type InternalMacroManagerOptions = {
  name: string
  runtime: Runtime
}

export function macroManager(
  options: InternalMacroManagerOptions
): MacroManager {
  const { name, runtime } = options

  return new MacroManagerImpl(name, runtime)
}

class MacroManagerImpl {
  private config?: ResolvedConfig
  private devServer?: ViteDevServer

  private plugins: Plugin[] = []

  private hooks: MacroProviderHooks[] = []

  constructor(name: string, private runtime: Runtime) {
    this.plugins.push(
      macroPlugin({
        name,
        runtime,
        hooks: {
          configResolved: (config) => this.handleConfigResolved(config),
          configureServer: (server) => this.handleConfigureServer(server),
          buildStart: () => this.handleBuildStart(),
        },
      })
    )
  }

  toPlugin() {
    return this.plugins.slice()
  }

  private get isRollup() {
    return !this.config
  }

  use(...sources: (MacroProvider | Plugin)[]) {
    sources.forEach((s) => this.add(s))
    return this
  }

  private add(p: MacroProvider | Plugin) {
    if (isMacroProvider(p)) this.addProvider(p)
    else if (isMacroPlugin(p)) this.addPlugin(p)
    else
      throw new Error(
        'the argument is neither a macro provider nor a macro plugin, ' +
          'please use defineMacroProvider() and defineMacroPlugin() to ' +
          'define providers and plugins.'
      )
  }

  private addProvider(provider: MacroProvider) {
    const { exports, options } = provider
    this.runtime.merge({ exports, options })
    this.hooks.push(provider.hooks)
  }

  private addPlugin(plugin: MacroPlugin) {
    const { exports, options } = plugin.__consume()
    this.runtime.merge({ exports, options })
    this.plugins.push(plugin)
  }

  async handleBuildStart() {
    if (this.isRollup)
      await Promise.all(this.hooks.map((h) => h.onRollupStart?.()))
    else
      await Promise.all(
        this.hooks.map((h) =>
          h.onViteStart?.(
            this.config!,
            this.devServer!,
            getDevServerHelper(this.devServer!)
          )
        )
      )
    await Promise.all(this.hooks.map((h) => h.onStart?.()))
  }

  handleConfigureServer(server: ViteDevServer) {
    this.devServer = server
    this.runtime.setDevMode()
  }

  handleConfigResolved(config: ResolvedConfig) {
    this.config = config
    this.runtime.setDevMode()
  }
}
