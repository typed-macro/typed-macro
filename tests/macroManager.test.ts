import { MacroPlugin, macroPlugin } from '@/macroPlugin'
import { isMacroManager, MacroManager, macroManager } from '@/macroManager'
import { macroProvider, MacroProvider } from '@/macroProvider'
import { withTempPath } from './testutils'

describe('macroManager() & isMacroManager()', () => {
  it('should work', () => {
    expect(
      isMacroManager(
        macroManager({
          name: 'test',
          runtimeOptions: {
            typeRenderer: {
              typesPath: '',
            },
            transformer: {},
          },
        })
      )
    ).toBe(true)
  })
})

// No need to test MacroManagerContext because
// MacroManager is just a simple wrapper of it
describe('MacroManager', () => {
  let _provider: MacroProvider
  let _plugin: MacroPlugin
  let _manager: MacroManager
  beforeEach(() => {
    _provider = macroProvider({
      id: 'provider',
      hooks: {},
      exports: {
        modules: {
          '@helper': `export function hello(msg) {console.log(msg)}`,
        },
        macros: {
          '@echo': [
            {
              name: 'echo',
              apply: ({ path }, { template }, { prependToBody }) => {
                prependToBody(
                  template.statements.ast(
                    `import { hello } from '@helper'; hello('world')`
                  )
                )
                path.remove()
              },
            },
          ],
        },
        types: {},
      },
    })
    _plugin = macroPlugin({
      name: 'plugin',
      hooks: {},
      runtimeOptions: {
        typeRenderer: { typesPath: '' },
        transformer: {},
      },
      exports: {
        modules: {},
        macros: {
          '@noop': [
            {
              name: 'noop',
              apply: ({ path }) => path.remove(),
            },
          ],
        },
        types: {},
      },
    })
    _manager = macroManager({
      name: 'test',
      runtimeOptions: {
        transformer: {},
        typeRenderer: { typesPath: '' },
      },
    })
  })

  it('should support consume providers/plugins', () => {
    expect(_manager.length).toBe(1)
    expect(() => _manager.use(_plugin)).not.toThrow()
    expect(() => _manager.use(_provider)).not.toThrow()
    expect(_manager.length).toBe(2)
  })

  it('should throw error when consume something not a provider/plugin', () => {
    expect(() => _manager.use({} as any)).toThrow()
  })

  it('should handle transform with macro', () => {
    _manager.use(_provider)
    const realPlugin = _manager[0]

    expect(
      realPlugin.transform!.call(
        null as any,
        `import { echo } from '@echo'`,
        ''
      )
    ).toBeUndefined()
    expect(
      realPlugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toMatchSnapshot()
  })

  it('should handle load properly', () => {
    _manager.use(_provider)
    const realPlugin = _manager[0]

    expect(realPlugin.load!.call(null as any, '@helper')).toMatchSnapshot()
    expect(realPlugin.load!.call(null as any, '@echo')).toMatchSnapshot()

    expect(realPlugin.load!.call(null as any, '@others')).toBeUndefined()
  })

  it('should handle resolveId properly', () => {
    _manager.use(_provider)
    const realPlugin = _manager[0]

    expect(realPlugin.resolveId!.call(null as any, '@helper', '', {})).toBe(
      '@helper'
    )
    expect(realPlugin.resolveId!.call(null as any, '@echo', '', {})).toBe(
      '@echo'
    )

    expect(
      realPlugin.resolveId!.call(null as any, '@others', '', {})
    ).toBeUndefined()
  })

  it('should handle hooks in buildStart() in expected order', (done) => {
    withTempPath('./a.d.ts', async (tempPath) => {
      const manager = macroManager({
        name: 'test',
        runtimeOptions: {
          transformer: {},
          typeRenderer: { typesPath: tempPath },
        },
      })

      const stack: string[] = []
      manager.use(
        macroProvider({
          id: 'provider',
          hooks: {
            onRollupStart: () => {
              stack.push('onRollupStart')
            },
            onViteStart: () => {
              stack.push('onViteStart')
            },
            onStart: () => {
              stack.push('onStart')
            },
          },
          exports: { modules: {}, macros: {}, types: {} },
        })
      )
      const realPlugin = manager[0]
      // rollup
      await realPlugin.buildStart!.call(null as any, {} as any)
      expect(stack).toEqual(['onRollupStart', 'onStart'])
      // vite
      stack.length = 0
      await realPlugin.configResolved!.call(null as any, {} as any)
      await realPlugin.buildStart!.call(null as any, {} as any)
      expect(stack).toEqual(['onViteStart', 'onStart'])
      done()
    })
  })
})
