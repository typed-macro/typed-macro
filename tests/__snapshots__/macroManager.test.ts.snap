// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MacroManager should handle load properly 1`] = `"export function hello(msg) {console.log(msg)}"`;

exports[`MacroManager should handle load properly 2`] = `"export {}"`;

exports[`MacroManager should handle transform with macro 1`] = `"import { hello } from '@helper';hello('world');"`;
